File: sdks/python/src/egma/monitoring.py
"""Export production spans from a LiveKit worker.

Select production by room name before connecting. Simulation rooms are
handled by ``egma.simulation`` and are ignored here.
"""

from __future__ import annotations

import logging

from livekit.agents import JobContext

from . import export
from .room import simulation_in

logger = logging.getLogger("egma")

VERB = "egma.monitor"
"""What this verb is called, for every sentence it has to say."""


def monitor(
    ctx: JobContext,
    *,
    endpoint: str | None = None,
    api_key: str | None = None,
) -> None:
    """Export this LiveKit worker's production spans to Egma.

    Call once before ``AgentSession.start``. ``endpoint`` and ``api_key``
    default to ``EGMA_URL`` and ``EGMA_API_KEY``. Repeated calls with the
    same settings reuse the exporter; job shutdown flushes buffered spans.

    Raise ``ValueError`` for invalid settings, unsupported LiveKit/provider
    configuration, or another job requesting different settings in this
    process. Returns without changes in simulation rooms.
    """

    # Simulation rooms export through simulation(), not production monitoring.
    # Log suppression because a production room using the reserved prefix
    # would otherwise lose its monitoring evidence without notice.
    simulation = simulation_in(ctx)
    if simulation is not None:
        logger.warning(
            "this job runs in the Egma simulation room %r, so its spans are "
            "not exported to production Monitoring: egma.simulation exports "
            "them under that simulation instead",
            simulation.named,
        )
        return

    export.install(ctx, verb=VERB, endpoint=endpoint, api_key=api_key)

File: sdks/livekit-js/src/monitoring.ts
import { type JobContext } from "@livekit/agents";

import { installExport, type ExportOptions } from "./export.ts";
import { SIMULATION_ROOM_PREFIX } from "./room.ts";

export const MONITOR_VERB = "egma.monitor";

export type MonitorOptions = ExportOptions;

/**
 * Export production spans to Egma. Call first in the job entrypoint, before
 * AgentSession.start. Repeated calls reuse the exporter; each job gets a final flush.
 *
 * Throws Error for invalid settings, unsupported LiveKit telemetry APIs, an unsafe
 * tracer provider, or different settings in a second job. Export failures do not
 * stop the agent. In simulation rooms, simulation() handles export instead.
 */
export function monitor(ctx: JobContext, options: MonitorOptions = {}): void {
  const roomName = ctx?.job?.room?.name;
  if (
    typeof roomName === "string" &&
    roomName.startsWith(SIMULATION_ROOM_PREFIX)
  ) {
    // Said out loud, because a room name is chosen by whoever mints the join
    // token: a production room named to look like a simulation would lose its
    // Monitoring record, and a dropped trace is evidence a customer cannot get
    // back.
    console.warn(
      `Egma: ${JSON.stringify(roomName)} is an Egma simulation room, so its spans are not exported to production Monitoring. egma.simulation exports them under that simulation instead.`,
    );
    return;
  }

  installExport(ctx, options, MONITOR_VERB, "");
}

export {
  projectKey,
  traceEndpoint,
  type ExistingTelemetry,
} from "./export.ts";
